### PR TITLE
feat: Add common UI patcher and GoGraph UI patch handler

### PR DESCRIPTION
**Pull request by lorentzfb**
This PR adds an alternative way of handling communication of graph patches to the UI. Two main files have been added: 

- src/core/ui/**applyPatch.ts**
> This file lives in the "core" and can be used of all UI implementations (goGraph, cyGraph, etc.). This common function is responsible for invoking the UI (via an IUiPatchHandler) when an assertion is processed. ~~It also deals with synchronisation logic (it may invoke multiple UI functions on a single assertion).~~

- src/goGraph/**uiPatchHandler.ts**
> An implementation of the IUiPatchHandler which is a required argument on the UiNegotiator. A patch handler should only be required to implement the IUiPatchHandler for everything to work (synchronisation is handled by **patchGraph**).

This approach simplifies the UI implementations and centralise common UI logic.

NOTE: src/goGraph/applyPatch.ts becomes obsolete for goGraph UI (not yet deleted).

### Sadly, this became a huge PR that covered more than was intended 😞 

Some other changes:

- Symbols for GoGraph use the 'geometryString' on a go.Shape instead of setting an SVG string as 'source' on a go.Picture (Only two symbols have been converted: Compressor_1 and Valve_Gate_O). CyGraph not affected.
- Go node symbol template tuned to position connectors perfectly 🎉 
- patchGraph prop assertions changed: relativePosition => connectorRelativePositon, direction => connectorDirection
- A symbolProvider can be injected as a prop on a UI implementation.

@toftalisk I challenge you: convert cyGraph to use a IUiPatchHandler 🚀

### CODE REVIEW

File: src/core/ui/uiNegotiator.ts [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+			default:
+				return;
+		}
+
+		if (action === 'add') this.ui.addConnectorProperty(assertion.target.id, assertion.target.node.id, propKey, assertion.value);
+		else this.ui.removeConnectorProperty(assertion.target.id, assertion.target.node.id, propKey);
+
+		// Side effects when connector 'name' property is set
+		if (propKey === 'name') {
```
Comment (toftalisk) [inline src/core/ui/uiNegotiator.ts:241]: We need to talk about this

File: src/goGraph/storybook/GoGraph.stories.tsx [hunk 2]
```diff
@@ -8,7 +8,11 @@ import { esd_turtle_border as esd_turtle } from './data/esd-turtle';
 //import * as TURTLE from './data-no-vc/SHUTDOWN_PRES';
 import { martinsTurtle } from './data/martins_verden_turtle';
 
-import { Assertion, GraphProperty, GraphSelection } from '../../core/types';
+import { lorentz_turtle } from './data/lorentz_ttl';
+
+//import lorentzTurtle from './data/lorentz.ttl?raw';
+
```
Comment (toftalisk) [inline src/goGraph/storybook/GoGraph.stories.tsx:14]:
```suggestion
```

File: src/goGraph/storybook/StoryWrapper.tsx [hunk 3]
```diff
@@ -29,7 +29,7 @@ export const StoryWrapper = ({ turtleString, layout, selectionEffect }: SparqlWr
 
 	function handleSelection(sel: GraphSelection): PropertyAssertion[] {
 		if (!selectionEffect) {
-			console.log('No selection handler for ', sel);
+			//console.log('No selection handler for ', sel);
```
Comment (toftalisk) [inline src/goGraph/storybook/StoryWrapper.tsx:32]:
```suggestion
```

File: src/goGraph/types/goNode.types.ts [hunk 4]
```diff
@@ -55,7 +62,7 @@ export type SymbolNodeData = BaseNodeData & {
 	height?: number;
 	width?: number;
 	angle?: number;
-	symConnectors?: NodeSymbolConnector[];
+	symbolConnectors?: NodeSymbolConnector[];
```
Comment (toftalisk) [inline src/goGraph/types/goNode.types.ts:65]: 👍

File: src/goGraph/uiPatchHandler.ts [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+};
+
+const edgePropMap: Record<keyof UiEdgePatchProperties, string> = {
+	color: 'color',
+	highlight: 'highlight',
+	highlightColor: 'highlightColor',
+	thickness: 'stroke',
+};
+
+export class GoJsPatchHandler implements IUiPatchHandler {
```
Comment (toftalisk) [inline src/goGraph/uiPatchHandler.ts:34]: This is class is awesome. Everything that is difficult here is related to gojs - api. All complex graph stuff is abstracted away before we get to this point. I think introducing IUiPatchHandler interface and UiNegotiator to divide responsibility was a great idea 👍 👍

**Review by toftalisk — COMMENTED** [review]
A lot of comments. Some can be addressed in this PR, but for some it is ok to fix in other, later PRs. I think we should merge this asap and continue to improve this idea.

I did not like that we needed to keep state in `UiNegotiator` as well, but since the interface outside it is clean it should be easy to fix .  Hopefully we will find a way around that.

I am missing some tests that should be added asap after merge.

Even though there was a lot of nitpicking here, I think the overall UiNegotiator idea was great and we are rewarded with a really easy interface against our Gui Implementations. I am really eager to test it out in cytoscape and continue to improve it. 

Need to test it a bit more before approve

File: src/core/types/graphModel.ts [hunk 6]
```diff
@@ -42,22 +42,24 @@ export type GraphMetadata = GraphElementBase & {
...
 	origin: Quad | GraphEdge;
 } & GraphVisualProps;
 
-export type GraphProperty = {
+export type GraphPropertyTarget = AbstractNode | GraphEdge;
+
+export type GraphProperty<TTarget extends GraphPropertyTarget> = {
 	type: 'property';
-	node: AbstractNode | GraphEdge;
+	target: TTarget;
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/types/graphModel.ts:62]: Introduced this to get type completion in ex.:

https://github.com/equinor/rdf-graph/blob/9f44d3b6ecc949033a3aa6e9060b953ccd030442/src/core/ui/uiNegotiator.ts#L252

File: src/core/ui/uiNegotiator.ts [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+	addEdgeProperty<P extends keyof UiEdgePatchProperties>(edgeId: string, prop: P, value: UiEdgePatchProperties[P]): void;
+	removeEdgeProperty<P extends keyof UiEdgePatchProperties>(edgeId: string, prop: P): void;
+
+	getNodeSymbol(id: string): UiNodeSymbol;
+
+	onBeforeApplyPatch?: () => void;
+	onAfterApplyPatch?: () => void;
+}
+
+// Future UiNegotiator options
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/ui/uiNegotiator.ts:84]: Any future options that can be set when using the negotiator in a UI impl. Dont have any examples of options yet, but I like the design part of it.

File: src/core/ui/uiNegotiator.ts [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+export type UiNegotiatorOptions = {};
+export const UiNegotiatorDefaultOptions = {} as const;
+
+/**  */
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/ui/uiNegotiator.ts:88]: Removed

File: src/core/ui/uiNegotiator.ts [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+				case 'connector':
+					this.patchConnector({ action, assertion });
+					break;
+				case 'property':
+					this.patchProperty({ action, assertion });
+					break;
+				default:
+					continue;
+			}
+			// MARTIN: Handy console logs for debug. We will remove when this class is finished.
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/ui/uiNegotiator.ts:120]: 💋

File: src/core/ui/uiNegotiator.ts [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+			// 		' - ' +
+			// 		(assertion.target as GraphPropertyTarget)?.type +
+			// 		' - ' +
+			// 		assertion.key +
+			// 		' - ' +
+			// 		JSON.stringify(assertion.value)
+			// );
+		}
+
+		// Invoke onAfterApplyPatch if defined
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/ui/uiNegotiator.ts:134]: sure. removed 😇

File: src/core/ui/UiNegotiatorNodeStore.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
+type NodeDataCache = {
+	symbolId: string | null;
+	connectors: Map<string, ConnectorData>;
+};
+
+type ConnectorData = {
+	name: string | null;
+};
+
+/** Store for node data used for UI synchronization  */
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/ui/UiNegotiatorNodeStore.ts:10]: Updated comment

File: src/core/ui/uiNegotiator.ts [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+	position: Point | 'Left' | 'Right' | 'Top' | 'Bottom';
+	/** The angle that the connector has to  */
+	normalDirection: number;
+}
+
+export interface UiNodeSymbol {
+	id: string;
+	width: number;
+	height: number;
+	//angle?: number;
```
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/ui/uiNegotiator.ts]: Uncommented

File: src/core/types/graphModel.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/types/graphModel.ts:62]: TTarget is a dotnet naming convention 😄

File: src/core/ui/uiNegotiator.ts (hunk 1 above)
Comment (lorentzfb) [reply src/core/ui/uiNegotiator.ts]: NA after last commit

File: src/goGraph/uiPatchHandler.ts (hunk 5 above)
Comment (lorentzfb) [reply src/goGraph/uiPatchHandler.ts:44]: We keep this:)

**Comment by lorentzfb** [comment]
> A lot of comments. Some can be addressed in this PR, but for some it is ok to fix in other, later PRs. I think we should merge this asap and continue to improve this idea.
> 
I think we are close to merging now!

> I did not like that we needed to keep state in `UiNegotiator` as well, but since the interface outside it is clean it should be easy to fix . Hopefully we will find a way around that.
> 

I've removed the state in UiNegotiator. Now its just a simple function that's loops through the generator and invokes the UI patch handler. As simple as it gets.

> I am missing some tests that should be added asap after merge.

Agree.
> 
> Even though there was a lot of nitpicking here, I think the overall UiNegotiator idea was great and we are rewarded with a really easy interface against our Gui Implementations. I am really eager to test it out in cytoscape and continue to improve it.
> 
> Need to test it a bit more before approve

Should be ready for testing now.

File: src/core/state/patchGraph.ts [hunk 13]
```diff
@@ -81,11 +82,20 @@ function removeId<T, K extends keyof T>(index: Map<string, T[]>, key: string, pr
...
+
 const dataProps = [
 	'symbolName',
 	'symbol',
+	'symbolGeometry',
 	'relativePosition',
 	'connectorName',
+	'connectorDirection',
+	'connectorRelativePosition',
+	//'connectorPosition',
```
Comment (toftalisk) [inline src/core/state/patchGraph.ts:98]:
```suggestion
```

File: src/core/types/graphModel.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (toftalisk) [reply src/core/types/graphModel.ts:62]: Aha 👍 convention is T-prefix for types?

File: src/core/ui/applyPatch.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+	highlight: boolean;
+	thickness: string;
+}
+
+export interface UiConnectorPatchProperties {
+	name: string;
+	color: string;
+	connectorRelativePosition: Point | 'Left' | 'Right' | 'Top' | 'Bottom';
+	/** The angle that the connector arm has to the node at default rotation */
+	connectorDirection: number;
```
Comment (toftalisk) [inline src/core/ui/applyPatch.ts:42]:
```suggestion
	connectorDirection: number; // The angle that the connector arm has to the node at default rotation
```

Minor, but find it more trivial to match code with comment this way

File: src/core/ui/applyPatch.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+export interface UiNodeSymbol {
+	id: string;
+	width: number;
+	height: number;
+	angle?: number;
+	svg?: string;
+	/** Symbol as single svg <path> element value */
+	geometry?: string;
```
Comment (toftalisk) [inline src/core/ui/applyPatch.ts:52]:
```suggestion
	geometry?: string; // Symbol as single svg <path> element value
```

File: src/core/ui/applyPatch.ts [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+/** Apply a GraphPatch to the UI using the IUiPatchHandler */
+export function applyPatch(graphPatch: GraphPatch, ui: IUiPatchHandler): void {
+	ui.onBeforeApplyPatch?.call(ui);
+
+	for (const { action, assertion } of graphPatch) {
+		// console.log(
+		// 	action + ' type=' + assertion.type + ' id=' + assertion.key + ' value="' + JSON.stringify(assertion.value) + '" targ=' + assertion.target
+		// );
```
Comment (toftalisk) [inline src/core/ui/applyPatch.ts:91]:
```suggestion
```

File: src/core/ui/applyPatch.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
+import { GraphPatch } from '../types';
```
Comment (toftalisk) [inline src/core/ui/applyPatch.ts:1]: Very nice and clean code in this file now 👍

File: src/goGraph/item-templates/position-port-item-template.ts [hunk 18]
```diff
@@ -64,11 +64,17 @@ export function createPositionPortItemTemplate(): go.Panel {
...
+				return new go.Point(pos.x, pos.y);
+			}),
+			//new go.Binding('position', 'relativePosition'),
+			// new go.Binding('alignment', 'relativePosition', (p) => {
+			// 	return new go.Spot(p.x, p.y);
+			// }),
+			// new go.Binding('position', 'portSize', (size, shape: go.Shape) => {
+			// 	const pos = shape.position;
+			// 	return new go.Point(pos.x - size / 2, pos.y - size / 2);
+			// }).ofModel(),
```
Comment (toftalisk) [inline src/goGraph/item-templates/position-port-item-template.ts:77]:
```suggestion
```

File: src/goGraph/node-templates/symbol-node-template.ts [hunk 19]
```diff
@@ -1,13 +1,13 @@
 import go from 'gojs';
 import { itemTemplateMap } from '../item-templates/item-templates-map';
```
Comment (toftalisk) [inline src/goGraph/node-templates/symbol-node-template.ts:2]: Fix unused stuff

File: src/goGraph/node-templates/symbol-node-template.ts [hunk 20]
```diff
@@ -29,64 +29,81 @@ export function createSymbolNodeTemplate(clickHandler?: ((e: go.InputEvent, this
...
+			// 		)
+			// 		.bind(
+			// 			new go.Binding('alignmentFocus', 'angle', (a: number) => {
+			// 				return getTopLabelAlignment(a);
+			// 			}).ofObject()
+			// 		)
+			// 		.bind(new go.Binding('text', 'label'))
+			// 		.bind(new go.Binding('stroke', 'uiTheme', ({ node }) => node.text).ofModel())
+			// )
+			// SYMBOL
```
Comment (toftalisk) [inline src/goGraph/node-templates/symbol-node-template.ts:56]:
```suggestion
```

File: src/goGraph/node-templates/symbol-node-template.ts [hunk 21]
```diff
@@ -29,64 +29,81 @@ export function createSymbolNodeTemplate(clickHandler?: ((e: go.InputEvent, this
...
-							.bind('itemArray', 'ports')
-							.bind('width')
-							.bind('height')
+			)
+			// CONNECTOR PANEL
+			.add(
+				new go.Panel(go.Panel.Position, {
+					//itemTemplateMap: itemTemplateMap,
+					//background: 'grey',
+					//opacity: 0.5,
```
Comment (toftalisk) [inline src/goGraph/node-templates/symbol-node-template.ts:75]:
```suggestion
```

File: src/goGraph/node-templates/symbol-node-template.ts [hunk 22]
```diff
@@ -29,64 +29,81 @@ export function createSymbolNodeTemplate(clickHandler?: ((e: go.InputEvent, this
...
 			.add(
-				new go.TextBlock('', {
-					alignment: go.Spot.TopCenter,
-					alignmentFocus: go.Spot.TopCenter,
-				})
-					.bind(new go.Binding('text', 'key').ofObject())
-					.bind(new go.Binding('angle', 'angle', (a) => -a).ofObject())
-					.bind(
-						new go.Binding('alignment', 'angle', (a: number) => {
-							return getTopLabelAlignment(a);
```
Comment (toftalisk) [inline src/goGraph/node-templates/symbol-node-template.ts:50]: Remove function as well

File: src/goGraph/storybook/GoGraph.stories.tsx [hunk 23]
```diff
@@ -8,7 +8,11 @@ import { esd_turtle_border as esd_turtle } from './data/esd-turtle';
 //import * as TURTLE from './data-no-vc/SHUTDOWN_PRES';
 import { martinsTurtle } from './data/martins_verden_turtle';
 
-import { Assertion, GraphProperty, GraphSelection } from '../../core/types';
+import { lorentz_turtle } from './data/lorentz_ttl';
+
+import { lorentz_turtle_2 } from './data/lorentz_ttl_2';
```
Comment (toftalisk) [inline src/goGraph/storybook/GoGraph.stories.tsx:13]: Remove unused import

File: src/goGraph/storybook/StoryWrapper.tsx [hunk 24]
```diff
@@ -8,13 +8,24 @@ import { useRdfActionReducer } from '../../core/state/useRdfState';
...
+import { UiNodeSymbol } from '../../core/ui/applyPatch';
+import { NodeSymbolToUiNodeSymbol } from '../../core/ui/defaultSymbolProvider';
 
 export type SparqlWrapperProps = {
 	turtleString: string;
 	layout: GoGraphLayout;
 	selectionEffect?: SelectionCallback;
 };
 
+function symbolProvider(id: string, rotation?: number): UiNodeSymbol | undefined {
```
Comment (toftalisk) [inline src/goGraph/storybook/StoryWrapper.tsx:21]: Remove rotation param or prefix it with `_` to remove warning

File: src/core/ui/uiNegotiator.ts [hunk 25]
```diff
@@ -1,6 +1,4 @@
-import { nodeTemplateKey } from '../mapper/predicates';
-import { Assertion, GraphConnector, GraphEdge, GraphNode, GraphPatch, GraphProperty, GraphPropertyTarget } from '../types';
-import { UiNegotiatorNodeStore } from './UiNegotiatorNodeStore';
```
Comment (toftalisk) [inline src/core/ui/uiNegotiator.ts]: ❤️

**Review by toftalisk — CHANGES_REQUESTED** [review]
This code is looking great! Really glad you managed to remove state from UiNegotiator.

I had some minor comments as always and I think it is easiest to resolve them before merge.

File: src/core/state/patchGraph.ts (hunk 13 above)
Comment (lorentzfb) [reply src/core/state/patchGraph.ts:98]: Already removed this locally :)

File: src/core/types/graphModel.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lorentzfb) [reply src/core/types/graphModel.ts:62]: Yes, "Type for Target"

File: src/core/ui/applyPatch.ts (hunk 14 above)
Comment (lorentzfb) [reply src/core/ui/applyPatch.ts:42]: Then we don't se this description when you hover the property in another file :(

File: src/core/ui/applyPatch.ts (hunk 15 above)
Comment (lorentzfb) [reply src/core/ui/applyPatch.ts:52]: Same as above..

File: src/core/ui/applyPatch.ts (hunk 16 above)
Comment (lorentzfb) [reply src/core/ui/applyPatch.ts:91]: Nice to have for debugging

File: src/goGraph/node-templates/symbol-node-template.ts (hunk 22 above)
Comment (lorentzfb) [reply src/goGraph/node-templates/symbol-node-template.ts:50]: This is WIP, will keep.

File: src/goGraph/storybook/GoGraph.stories.tsx (hunk 23 above)
Comment (lorentzfb) [reply src/goGraph/storybook/GoGraph.stories.tsx:13]: They are used???

File: src/goGraph/storybook/StoryWrapper.tsx (hunk 24 above)
Comment (lorentzfb) [reply src/goGraph/storybook/StoryWrapper.tsx]: Fixed

File: src/goGraph/node-templates/symbol-node-template.ts (hunk 21 above)
Comment (lorentzfb) [reply src/goGraph/node-templates/symbol-node-template.ts:75]: WIP

File: src/goGraph/item-templates/position-port-item-template.ts (hunk 18 above)
Comment (lorentzfb) [reply src/goGraph/item-templates/position-port-item-template.ts:77]: ok

**Review by toftalisk — APPROVED** [review]
Awesome 🎉

**Pull request merged**